Misc. minor fixes

Daniel-Cortez opened the pull request:
**What this PR does / why we need it**:

Fixes for various minor bugs that weren't worth making a separate PR for each of them:

* Fixed invalid formula for calculating array size (in characters) in function `hier1()`.
  The cell size in bits was hardcoded as `32` (should be `sizeof(cell)*8`), which could lead to generation of invalid code for array boundary checks if the Pawn toolkit was configured with 16- or 64-bit cell size.
  Of course, this bug doesn't affect SA-MP (which is 32-bit), but I think it's still good to have this fix here.

* Fixed invalid type cast in `funcstub()`.
  See https://github.com/pawn-lang/compiler/issues/296#issuecomment-450643026.

* Fixed OOB array access in `encode_cell()`.
  This one seems to originate from a commit that I made a year and a half ago (sorry!)

* Fixed memory leaks in `plungequalifiedfile()`.
  Special thanks goes to @YashasSamaga for [reporting the bug, as well as providing a fix](https://github.com/pawn-lang/compiler/issues/296#issuecomment-449903004), though I slightly changed that fix in the following ways:
  * Moved `free(path)` before `error(103)`, because `path` won't be freed otherwise. This is because on fatal errors (1xx) `error()` returns (via `longjmp()`) directly to an error buffer set in `pc_compile`, so the code after such `error()` call is never executed.
  * For the same reason I removed two `return`s, as they are pretty much useless after `error(<fatal error code>)`.

* Removed `NULL` pointer checks before the uses of function `free()`.
  [`free()` does nothing if the pointer is `NULL`](https://en.cppreference.com/w/c/memory/free), so any `NULL` checks before the use of `free()` are redundant.

* Fixed invalid uses of `realloc()` in several places.
  The old pointer could have got leaked if the function failed to resize the old memory block or allocate a new one.

* Simplified the code in function `pushstk()` by replacing a group of calls to `malloc()`/`memcpy()`/`free()` with one `realloc()`.

* Removed function `do_switch()` (`sc6.c`).
  Apparently this function was a duplicate of `do_jump()`; it didn't implement any extra checks, so its only difference from `do_jump()` was in its name.

* Simplified string emptiness checks.
  Using `strlen(<string>)==0` to identify empty strings is inefficient because if the string is not empty `strlen()` traverses through the whole string to get its length.

* Fixed potential `NULL` pointer dereferences in functions `sc_path()` and `setconfig()`.

* Fixed incorrect error message when compact encoding is ineffective (#425).

* Other minor adjustments.

**Which issue(s) this PR fixes**:

Fixes #425

**What kind of pull this is**:

* [x] A Bug Fix
* [ ] A New Feature
* [ ] Some repository meta (documentation, etc)
* [ ] Other

**Additional Documentation**:

Y-Less commented:
Looks fine to me, but only from a visual readthrough.

>       * For the same reason I removed two `return`s, as they are pretty much useless after `error(<fatal error code>)`.
> 

I couldn't find those returns.  Did you change this part?  I'd suggest not removing returns after fatal errors, since the c compiler doesn't know that `error` might not return (it could be marked as `noreturn`, but it does in some circumstances).  Thus the return will cut off certain code paths and allow the compiler compiler to do more when compiling the compiler.

Daniel-Cortez commented:
> I couldn't find those returns. Did you change this part?

No, I meant that I removed them from the code @YashasSamaga added in their original fix.

> I'd suggest not removing returns after fatal errors, since the c compiler doesn't know that error might not return (it could be marked as noreturn, but it does in some circumstances). Thus the return will cut off certain code paths and allow the compiler compiler to do more when compiling the compiler.

Makes sense, but I removed those returns not only because they would be unreachable, but also to be more in-line with the existing code (in all the other code there are usually no returns after fatal errors).

Y-Less commented:
Oh I see, then fair enough.

Daniel-Cortez commented:
@Zeex Done.

Zeex commented:
Thanks, looks good
